src/pdn/src/straps.cpp

```diff
@@ -413,6 +414,70 @@ FollowPins::FollowPins(Grid* grid, odb::dbTechLayer* layer, int width)
   }
 }
 
+std::unordered_map<odb::dbRow*, RowEndInsts> FollowPins::findRowEndInsts()
+{
+  std::unordered_map<odb::dbRow*, RowEndInsts> result;
+  auto rows = getDomain()->getRows();
+
+  std::unordered_map<int, odb::dbRow*> y_to_row;
+  for (auto row : rows) {
+    y_to_row[row->getBBox().yMin()] = row;
+  }
+
+  for (auto inst : getGrid()->getBlock()->getInsts()) {
+    if (!inst->isFixed()
+        || (!inst->isEndCap()
+            && inst->getMaster()->getType()
+                   != odb::dbMasterType::CORE_WELLTAP)) {
+      continue;
+    }
+    auto box = inst->getBBox();
+    auto found_row = y_to_row.find(box->yMin());
+    if (found_row == y_to_row.end()) {
+      continue;
+    }
+    odb::dbRow* row = found_row->second;
+    if (box->xMin() == row->getBBox().xMin()) {
+      result[row].left = inst;
+    }
+    if (box->xMax() == row->getBBox().xMax()) {
+      result[row].right = inst;
+    }
```

Comment:
To check I've understood the cut row issue correctly and can implement fixes, also for the above comment, when a row is cut by a macro, there will be two rows either side of it on the same y-location but with different x bounding boxes?

Importantly, what I'd like to confirm is there's never a case where you have a single dbRow with a "hole" inside of it?

Reply:
I've fixed cut rows by indexing on lower-left and lower-right points instead of just on y-coordinates. This seems to work on a local test.

 I haven't committed the test yet, because although it works well enough to check that the PDN stops at the right point, the endcap/corner insertion around macros isn't really playing well with the ring design of endcap cells that's causing problems here, so it would be quite a strange looking test floorplan with short circuits elsewhere.